projects/TR3D/tr3d/rotated_iou_loss.py

```diff
@@ -0,0 +1,149 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from typing import Optional
+
+import torch
+from mmcv.ops.diff_iou_rotated import box2corners, oriented_box_intersection_2d
+from mmdet.models.losses.utils import weighted_loss
+from torch import Tensor
+from torch import nn as nn
+
+from mmdet3d.models import rotated_iou_3d_loss
+from mmdet3d.registry import MODELS
+
+
+def diff_diou_rotated_3d(box3d1: Tensor, box3d2: Tensor) -> Tensor:
+    """Calculate differentiable DIoU of rotated 3d boxes.
+
+    Args:
+        box3d1 (Tensor): (B, N, 3+3+1) First box (x,y,z,w,h,l,alpha).
+        box3d2 (Tensor): (B, N, 3+3+1) Second box (x,y,z,w,h,l,alpha).
+    Returns:
+        Tensor: (B, N) DIoU.
+    """
+    box1 = box3d1[..., [0, 1, 3, 4, 6]]
+    box2 = box3d2[..., [0, 1, 3, 4, 6]]
+    corners1 = box2corners(box1)
+    corners2 = box2corners(box2)
+    intersection, _ = oriented_box_intersection_2d(corners1, corners2)
+    zmax1 = box3d1[..., 2] + box3d1[..., 5] * 0.5
+    zmin1 = box3d1[..., 2] - box3d1[..., 5] * 0.5
+    zmax2 = box3d2[..., 2] + box3d2[..., 5] * 0.5
+    zmin2 = box3d2[..., 2] - box3d2[..., 5] * 0.5
+    z_overlap = (torch.min(zmax1, zmax2) -
+                 torch.max(zmin1, zmin2)).clamp_(min=0.)
+    intersection_3d = intersection * z_overlap
+    volume1 = box3d1[..., 3] * box3d1[..., 4] * box3d1[..., 5]
+    volume2 = box3d2[..., 3] * box3d2[..., 4] * box3d2[..., 5]
+    union_3d = volume1 + volume2 - intersection_3d
```

Comment:
Line 23-37 seems have been implemented in `diff_iou_rotated_3d` [here](https://github.com/open-mmlab/mmcv/blob/2.x/mmcv/ops/diff_iou_rotated.py#L276). We may directly use this function for simplify.

Reply:
But we need lines 23-26 and 28-31 after 37 line

Reply:
Yeah, you are right. It's my fault. Thanks for your reminder.